DllImportGenerator/DllImportGenerator/DllImportGenerator.cs

```diff
@@ -293,96 +312,193 @@ private DllImportStub.GeneratedDllImportData ProcessGeneratedDllImportAttribute(
                     default:
                         Debug.Fail($"An unknown member was found on {GeneratedDllImport}");
                         continue;
-                    case nameof(DllImportStub.GeneratedDllImportData.BestFitMapping):
-                        stubDllImportData.BestFitMapping = (bool)namedArg.Value.Value!;
-                        stubDllImportData.IsUserDefined |= DllImportStub.DllImportMember.BestFitMapping;
+                    case nameof(GeneratedDllImportData.BestFitMapping):
+                        stubDllImportData = stubDllImportData with
```

Comment:
All this `with` logic is simply continually making copies. What is the win here? I don't see the value as opposed to just updating an instance before it leaves scope. If the language can't express that we shouldn't defer to multiple copies being made.

Reply:
The reason for switching to `with` expressions is to allow us to update the instance while using `init` setters to enforce that once the object is constructed, it is immutable.

I can change `GeneratedDllImportData` to be a `record struct` so the copies can be non-allocating and possibly be elided by the JIT completely. Alternatively, I can make an alternative data type that enforces the immutability.

Reply:
Is `GeneratedDllImportData` a public type? If so, then let's try the `record struct` method. If not, I don't think we should constrain the implementation based on a limitation of the `init` feature. The methodology here can be constrained to our usage and policy without much issue.

Reply:
At the moment it is, but it likely won't need to be once we finish the extensibility work.

My bigger concern is around ensuring that we don't break assumptions in how the incremental pipeline works. Since `GeneratedDllImportData` instances are passed between the phases of the incremental generator, they are cached by Roslyn between runs. As a result, we can end up incorrectly reusing cached results if we accidentally add code that changes the value of one of these instances. This isn't too bad right now since incremental results are only used in the IDE, but with https://github.com/dotnet/roslyn/pull/55023, this type of bug is much scarier.

I personally don't like relying on implicit policy when compiler enforcement is possible. It usually ends up in someone not noticing a misuse of a type and a hard-to-track-down bug.

I'll rewrite this to use locals internally to remove all of the copies and construct the `GeneratedDllImportData` object at the end.